### PR TITLE
[CI] Link checker GitHub workflow

### PR DESCRIPTION
**Pull request by pmm-sumo**
**Description:**

Adds link checking (for changed files)

This copies the solution already existing in core: https://github.com/open-telemetry/opentelemetry-collector/blob/main/.github/workflows/check-links.yaml

**Link to tracking Issue:** #10162 

**Testing:** Tested the change on my repo with sample content

**Documentation:** N/A

### CODE REVIEW

**Comment by pmm-sumo** [comment]
@codeboten @TylerHelmuth following our discussion, here's the link checker workflow

File: .github/workflows/check-links.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+      - name: Install markdown-link-check
+        run: npm install -g markdown-link-check
+
+      - name: Run markdown-link-check
+        run: |
+          markdown-link-check \
+            --verbose \
+            --config .github/workflows/check_links_config.json \
+            ${{needs.changedfiles.outputs.md}}
```
Comment (TylerHelmuth) [inline .github/workflows/check-links.yaml:44]:
```suggestion
            ${{needs.changedfiles.outputs.md}}
            
```

File: .github/workflows/check-links.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
+name: check-links
+on:
+  push:
+    branches: [ main ]
+    paths-ignore:
+      - 'cmd/builder/**'
+  pull_request:
+    paths-ignore:
+      - 'cmd/builder/**'
```
Comment (TylerHelmuth) [inline .github/workflows/check-links.yaml:9]: Contrib doesn't have `cmd/builder`.  Do we have an equivalent we need to ignore or can we remove this?

**Review by TylerHelmuth — APPROVED** [review]
LGTM, just 2 small questions.  Also, have you run this to check if there are any existing broken links?  If there is anything broken we should fix them as part of this PR so that the workflow doesn't start failing everyone.

**Comment by pmm-sumo** [comment]
> Also, have you run this to check if there are any existing broken links?

I did and there's a bunch of these (including CHANGELOG.md). I think this could be done separately. Some of them are fun ones (like old references in CHANGELOG to items that no longer exist) - should we retroactively update these or keep them?

**Comment by TylerHelmuth** [comment]
> > Also, have you run this to check if there are any existing broken links?
> 
> I did and there's a bunch of these (including CHANGELOG.md). I think this could be fixed separately. Some of them are fun ones (like old references in CHANGELOG to items that no longer exist) - should we retroactively update these or keep them?

Uh oh lol does the workflow work like a warning or will it start throwing up red Xs on each PR?

**Comment by pmm-sumo** [comment]
> Uh oh lol I feel like those dead links should stay dead. Maybe we should exclude changelog? Does the workflow work like a warning or will it start throwing up red Xs on each PR?

So the workflow checks only the changed files, which I think is a reasonable compromise and saves us from fixing ton of links

**Comment by TylerHelmuth** [comment]
> So the workflow checks only the changed files, which I think is a reasonable compromise and saves us from fixing ton of links

Good point, thats a good compromise.  What about changelog tho?  That is changed with each PR.

**Comment by pmm-sumo** [comment]
> > So the workflow checks only the changed files, which I think is a reasonable compromise and saves us from fixing ton of links
> 
> Good point, thats a good compromise. What about changelog tho? That is changed with each PR.

To be more accurate, I think it only looks at the diff, so only changed lines

**Comment by TylerHelmuth** [comment]
> To be more accurate, I think it only looks at the diff, so only changed lines

You're right: `name=md::$(git diff --name-only --diff-filter=ACMRTUXB origin/${{ github.event.pull_request.base.ref }} ${{ github.event.pull_request.head.sha }} | grep .md$ | xargs)"`

**Comment by pmm-sumo** [comment]
Well, I tested and it looks at the whole contents of the checked file, so I guess we should fix CHANGELOG.md :)

**Comment by TylerHelmuth** [comment]
> Well, I tested and it looks at the whole contents of the checked file, so I guess we should fix CHANGELOG.md :)

What are you thinking about the dead links?

**Comment by pmm-sumo** [comment]
> > Well, I tested and it looks at the whole contents of the checked file, so I guess we should fix CHANGELOG.md :)
> 
> What are you thinking about the dead links?

There are just 3 of them in the CHANGELOG.md, I think we can edit those and probably just pointing to a tag will fix them (we should also make sure that tag is used whenever links are posted):

```
  [✖] https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/attributeprocessor → Status: 404
  [✖] https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/contrib/jmx-metrics/README.md → Status: 404
  [✖] https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/stackdriverexporter/README.md → Status: 404
  ```

**Comment by TylerHelmuth** [comment]
> There are just 3 of them in the CHANGELOG.md, I think we can edit those and probably just pointing to a tag will fix them (we should also make sure that tag is used whenever links are posted):
> 
> ```
>   [✖] https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/attributeprocessor → Status: 404
>   [✖] https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/contrib/jmx-metrics/README.md → Status: 404
>   [✖] https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/stackdriverexporter/README.md → Status: 404
> ```

https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/attributeprocessor can probably be updated to https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/attributesprocessor

File: .github/workflows/check-links.yaml (hunk 2 above)
Comment (pmm-sumo) [reply .github/workflows/check-links.yaml]: Removed

**Review by codeboten — APPROVED** [review]

**Pull request merged**